From 5719b6d4f10b3657d5c04836ee5b8a074bb98a87 Mon Sep 17 00:00:00 2001
From: Firoz Khan <firozk@codeaurora.org>
Date: Sun, 1 Jul 2018 11:44:33 +0530
Subject: [PATCH 1/3] arch: ia64: add system call generation script

TODO:
- Support syscall.S, which extracted from entry.S file
- Remove "_64" unistd_64.h syscall_64.h
---
 arch/ia64/Makefile                      |   3 +
 arch/ia64/include/asm/Kbuild            |   1 +
 arch/ia64/include/uapi/asm/Kbuild       |   1 +
 arch/ia64/include/uapi/asm/unistd.h     | 329 +------------------------------
 arch/ia64/kernel/Makefile               |   2 +-
 arch/ia64/kernel/entry.S                | 332 -------------------------------
 arch/ia64/kernel/syscall.S              |  11 ++
 arch/ia64/kernel/syscalls/Makefile      |  37 ++++
 arch/ia64/kernel/syscalls/syscall.tbl   | 334 ++++++++++++++++++++++++++++++++
 arch/ia64/kernel/syscalls/syscallhdr.sh |  32 +++
 arch/ia64/kernel/syscalls/syscalltbl.sh |  29 +++
 11 files changed, 450 insertions(+), 661 deletions(-)
 create mode 100644 arch/ia64/kernel/syscall.S
 create mode 100644 arch/ia64/kernel/syscalls/Makefile
 create mode 100644 arch/ia64/kernel/syscalls/syscall.tbl
 create mode 100644 arch/ia64/kernel/syscalls/syscallhdr.sh
 create mode 100644 arch/ia64/kernel/syscalls/syscalltbl.sh

diff --git a/arch/ia64/Makefile b/arch/ia64/Makefile
index 45f5980..320d86f 100644
--- a/arch/ia64/Makefile
+++ b/arch/ia64/Makefile
@@ -80,6 +80,9 @@ unwcheck: vmlinux
 archclean:
 	$(Q)$(MAKE) $(clean)=$(boot)
 
+archheaders:
+	$(Q)$(MAKE) $(build)=arch/ia64/kernel/syscalls all
+
 CLEAN_FILES += vmlinux.gz bootloader
 
 boot:	lib/lib.a vmlinux
diff --git a/arch/ia64/include/asm/Kbuild b/arch/ia64/include/asm/Kbuild
index 557bbc8..cc7addd 100644
--- a/arch/ia64/include/asm/Kbuild
+++ b/arch/ia64/include/asm/Kbuild
@@ -7,3 +7,4 @@ generic-y += preempt.h
 generic-y += trace_clock.h
 generic-y += vtime.h
 generic-y += word-at-a-time.h
+generic-y += syscall_64.h
\ No newline at end of file
diff --git a/arch/ia64/include/uapi/asm/Kbuild b/arch/ia64/include/uapi/asm/Kbuild
index 3982e67..75b64e1 100644
--- a/arch/ia64/include/uapi/asm/Kbuild
+++ b/arch/ia64/include/uapi/asm/Kbuild
@@ -8,3 +8,4 @@ generic-y += msgbuf.h
 generic-y += poll.h
 generic-y += sembuf.h
 generic-y += shmbuf.h
+generic-y += unistd_64.h
\ No newline at end of file
diff --git a/arch/ia64/include/uapi/asm/unistd.h b/arch/ia64/include/uapi/asm/unistd.h
index 5fe71d4..e270b73 100644
--- a/arch/ia64/include/uapi/asm/unistd.h
+++ b/arch/ia64/include/uapi/asm/unistd.h
@@ -8,337 +8,10 @@
 #ifndef _UAPI_ASM_IA64_UNISTD_H
 #define _UAPI_ASM_IA64_UNISTD_H
 
-
 #include <asm/break.h>
 
 #define __BREAK_SYSCALL			__IA64_BREAK_SYSCALL
 
-#define __NR_ni_syscall			1024
-#define __NR_exit			1025
-#define __NR_read			1026
-#define __NR_write			1027
-#define __NR_open			1028
-#define __NR_close			1029
-#define __NR_creat			1030
-#define __NR_link			1031
-#define __NR_unlink			1032
-#define __NR_execve			1033
-#define __NR_chdir			1034
-#define __NR_fchdir			1035
-#define __NR_utimes			1036
-#define __NR_mknod			1037
-#define __NR_chmod			1038
-#define __NR_chown			1039
-#define __NR_lseek			1040
-#define __NR_getpid			1041
-#define __NR_getppid			1042
-#define __NR_mount			1043
-#define __NR_umount			1044
-#define __NR_setuid			1045
-#define __NR_getuid			1046
-#define __NR_geteuid			1047
-#define __NR_ptrace			1048
-#define __NR_access			1049
-#define __NR_sync			1050
-#define __NR_fsync			1051
-#define __NR_fdatasync			1052
-#define __NR_kill			1053
-#define __NR_rename			1054
-#define __NR_mkdir			1055
-#define __NR_rmdir			1056
-#define __NR_dup			1057
-#define __NR_pipe			1058
-#define __NR_times			1059
-#define __NR_brk			1060
-#define __NR_setgid			1061
-#define __NR_getgid			1062
-#define __NR_getegid			1063
-#define __NR_acct			1064
-#define __NR_ioctl			1065
-#define __NR_fcntl			1066
-#define __NR_umask			1067
-#define __NR_chroot			1068
-#define __NR_ustat			1069
-#define __NR_dup2			1070
-#define __NR_setreuid			1071
-#define __NR_setregid			1072
-#define __NR_getresuid			1073
-#define __NR_setresuid			1074
-#define __NR_getresgid			1075
-#define __NR_setresgid			1076
-#define __NR_getgroups			1077
-#define __NR_setgroups			1078
-#define __NR_getpgid			1079
-#define __NR_setpgid			1080
-#define __NR_setsid			1081
-#define __NR_getsid			1082
-#define __NR_sethostname		1083
-#define __NR_setrlimit			1084
-#define __NR_getrlimit			1085
-#define __NR_getrusage			1086
-#define __NR_gettimeofday		1087
-#define __NR_settimeofday		1088
-#define __NR_select			1089
-#define __NR_poll			1090
-#define __NR_symlink			1091
-#define __NR_readlink			1092
-#define __NR_uselib			1093
-#define __NR_swapon			1094
-#define __NR_swapoff			1095
-#define __NR_reboot			1096
-#define __NR_truncate			1097
-#define __NR_ftruncate			1098
-#define __NR_fchmod			1099
-#define __NR_fchown			1100
-#define __NR_getpriority		1101
-#define __NR_setpriority		1102
-#define __NR_statfs			1103
-#define __NR_fstatfs			1104
-#define __NR_gettid			1105
-#define __NR_semget			1106
-#define __NR_semop			1107
-#define __NR_semctl			1108
-#define __NR_msgget			1109
-#define __NR_msgsnd			1110
-#define __NR_msgrcv			1111
-#define __NR_msgctl			1112
-#define __NR_shmget			1113
-#define __NR_shmat			1114
-#define __NR_shmdt			1115
-#define __NR_shmctl			1116
-/* also known as klogctl() in GNU libc: */
-#define __NR_syslog			1117
-#define __NR_setitimer			1118
-#define __NR_getitimer			1119
-/* 1120 was __NR_old_stat */
-/* 1121 was __NR_old_lstat */
-/* 1122 was __NR_old_fstat */
-#define __NR_vhangup			1123
-#define __NR_lchown			1124
-#define __NR_remap_file_pages		1125
-#define __NR_wait4			1126
-#define __NR_sysinfo			1127
-#define __NR_clone			1128
-#define __NR_setdomainname		1129
-#define __NR_uname			1130
-#define __NR_adjtimex			1131
-/* 1132 was __NR_create_module */
-#define __NR_init_module		1133
-#define __NR_delete_module		1134
-/* 1135 was __NR_get_kernel_syms */
-/* 1136 was __NR_query_module */
-#define __NR_quotactl			1137
-#define __NR_bdflush			1138
-#define __NR_sysfs			1139
-#define __NR_personality		1140
-#define __NR_afs_syscall		1141
-#define __NR_setfsuid			1142
-#define __NR_setfsgid			1143
-#define __NR_getdents			1144
-#define __NR_flock			1145
-#define __NR_readv			1146
-#define __NR_writev			1147
-#define __NR_pread64			1148
-#define __NR_pwrite64			1149
-#define __NR__sysctl			1150
-#define __NR_mmap			1151
-#define __NR_munmap			1152
-#define __NR_mlock			1153
-#define __NR_mlockall			1154
-#define __NR_mprotect			1155
-#define __NR_mremap			1156
-#define __NR_msync			1157
-#define __NR_munlock			1158
-#define __NR_munlockall			1159
-#define __NR_sched_getparam		1160
-#define __NR_sched_setparam		1161
-#define __NR_sched_getscheduler		1162
-#define __NR_sched_setscheduler		1163
-#define __NR_sched_yield		1164
-#define __NR_sched_get_priority_max	1165
-#define __NR_sched_get_priority_min	1166
-#define __NR_sched_rr_get_interval	1167
-#define __NR_nanosleep			1168
-#define __NR_nfsservctl			1169
-#define __NR_prctl			1170
-/* 1171 is reserved for backwards compatibility with old __NR_getpagesize */
-#define __NR_mmap2			1172
-#define __NR_pciconfig_read		1173
-#define __NR_pciconfig_write		1174
-#define __NR_perfmonctl			1175
-#define __NR_sigaltstack		1176
-#define __NR_rt_sigaction		1177
-#define __NR_rt_sigpending		1178
-#define __NR_rt_sigprocmask		1179
-#define __NR_rt_sigqueueinfo		1180
-#define __NR_rt_sigreturn		1181
-#define __NR_rt_sigsuspend		1182
-#define __NR_rt_sigtimedwait		1183
-#define __NR_getcwd			1184
-#define __NR_capget			1185
-#define __NR_capset			1186
-#define __NR_sendfile			1187
-#define __NR_getpmsg			1188
-#define __NR_putpmsg			1189
-#define __NR_socket			1190
-#define __NR_bind			1191
-#define __NR_connect			1192
-#define __NR_listen			1193
-#define __NR_accept			1194
-#define __NR_getsockname		1195
-#define __NR_getpeername		1196
-#define __NR_socketpair			1197
-#define __NR_send			1198
-#define __NR_sendto			1199
-#define __NR_recv			1200
-#define __NR_recvfrom			1201
-#define __NR_shutdown			1202
-#define __NR_setsockopt			1203
-#define __NR_getsockopt			1204
-#define __NR_sendmsg			1205
-#define __NR_recvmsg			1206
-#define __NR_pivot_root			1207
-#define __NR_mincore			1208
-#define __NR_madvise			1209
-#define __NR_stat			1210
-#define __NR_lstat			1211
-#define __NR_fstat			1212
-#define __NR_clone2			1213
-#define __NR_getdents64			1214
-#define __NR_getunwind			1215
-#define __NR_readahead			1216
-#define __NR_setxattr			1217
-#define __NR_lsetxattr			1218
-#define __NR_fsetxattr			1219
-#define __NR_getxattr			1220
-#define __NR_lgetxattr			1221
-#define __NR_fgetxattr			1222
-#define __NR_listxattr			1223
-#define __NR_llistxattr			1224
-#define __NR_flistxattr			1225
-#define __NR_removexattr		1226
-#define __NR_lremovexattr		1227
-#define __NR_fremovexattr		1228
-#define __NR_tkill			1229
-#define __NR_futex			1230
-#define __NR_sched_setaffinity		1231
-#define __NR_sched_getaffinity		1232
-#define __NR_set_tid_address		1233
-#define __NR_fadvise64			1234
-#define __NR_tgkill			1235
-#define __NR_exit_group			1236
-#define __NR_lookup_dcookie		1237
-#define __NR_io_setup			1238
-#define __NR_io_destroy			1239
-#define __NR_io_getevents		1240
-#define __NR_io_submit			1241
-#define __NR_io_cancel			1242
-#define __NR_epoll_create		1243
-#define __NR_epoll_ctl			1244
-#define __NR_epoll_wait			1245
-#define __NR_restart_syscall		1246
-#define __NR_semtimedop			1247
-#define __NR_timer_create		1248
-#define __NR_timer_settime		1249
-#define __NR_timer_gettime		1250
-#define __NR_timer_getoverrun		1251
-#define __NR_timer_delete		1252
-#define __NR_clock_settime		1253
-#define __NR_clock_gettime		1254
-#define __NR_clock_getres		1255
-#define __NR_clock_nanosleep		1256
-#define __NR_fstatfs64			1257
-#define __NR_statfs64			1258
-#define __NR_mbind			1259
-#define __NR_get_mempolicy		1260
-#define __NR_set_mempolicy		1261
-#define __NR_mq_open			1262
-#define __NR_mq_unlink			1263
-#define __NR_mq_timedsend		1264
-#define __NR_mq_timedreceive		1265
-#define __NR_mq_notify			1266
-#define __NR_mq_getsetattr		1267
-#define __NR_kexec_load			1268
-#define __NR_vserver			1269
-#define __NR_waitid			1270
-#define __NR_add_key			1271
-#define __NR_request_key		1272
-#define __NR_keyctl			1273
-#define __NR_ioprio_set			1274
-#define __NR_ioprio_get			1275
-#define __NR_move_pages			1276
-#define __NR_inotify_init		1277
-#define __NR_inotify_add_watch		1278
-#define __NR_inotify_rm_watch		1279
-#define __NR_migrate_pages		1280
-#define __NR_openat			1281
-#define __NR_mkdirat			1282
-#define __NR_mknodat			1283
-#define __NR_fchownat			1284
-#define __NR_futimesat			1285
-#define __NR_newfstatat			1286
-#define __NR_unlinkat			1287
-#define __NR_renameat			1288
-#define __NR_linkat			1289
-#define __NR_symlinkat			1290
-#define __NR_readlinkat			1291
-#define __NR_fchmodat			1292
-#define __NR_faccessat			1293
-#define __NR_pselect6			1294
-#define __NR_ppoll			1295
-#define __NR_unshare			1296
-#define __NR_splice			1297
-#define __NR_set_robust_list		1298
-#define __NR_get_robust_list		1299
-#define __NR_sync_file_range		1300
-#define __NR_tee			1301
-#define __NR_vmsplice			1302
-#define __NR_fallocate			1303
-#define __NR_getcpu			1304
-#define __NR_epoll_pwait		1305
-#define __NR_utimensat			1306
-#define __NR_signalfd			1307
-#define __NR_timerfd			1308
-#define __NR_eventfd			1309
-#define __NR_timerfd_create		1310
-#define __NR_timerfd_settime		1311
-#define __NR_timerfd_gettime		1312
-#define __NR_signalfd4			1313
-#define __NR_eventfd2			1314
-#define __NR_epoll_create1		1315
-#define __NR_dup3			1316
-#define __NR_pipe2			1317
-#define __NR_inotify_init1		1318
-#define __NR_preadv			1319
-#define __NR_pwritev			1320
-#define __NR_rt_tgsigqueueinfo		1321
-#define __NR_recvmmsg			1322
-#define __NR_fanotify_init		1323
-#define __NR_fanotify_mark		1324
-#define __NR_prlimit64			1325
-#define __NR_name_to_handle_at		1326
-#define __NR_open_by_handle_at  	1327
-#define __NR_clock_adjtime		1328
-#define __NR_syncfs			1329
-#define __NR_setns			1330
-#define __NR_sendmmsg			1331
-#define __NR_process_vm_readv		1332
-#define __NR_process_vm_writev		1333
-#define __NR_accept4			1334
-#define __NR_finit_module		1335
-#define __NR_sched_setattr		1336
-#define __NR_sched_getattr		1337
-#define __NR_renameat2			1338
-#define __NR_getrandom			1339
-#define __NR_memfd_create		1340
-#define __NR_bpf			1341
-#define __NR_execveat			1342
-#define __NR_userfaultfd		1343
-#define __NR_membarrier			1344
-#define __NR_kcmp			1345
-#define __NR_mlock2			1346
-#define __NR_copy_file_range		1347
-#define __NR_preadv2			1348
-#define __NR_pwritev2			1349
+#include <asm/unistd_64.h>
 
 #endif /* _UAPI_ASM_IA64_UNISTD_H */
diff --git a/arch/ia64/kernel/Makefile b/arch/ia64/kernel/Makefile
index 498f3da..de0d7e6 100644
--- a/arch/ia64/kernel/Makefile
+++ b/arch/ia64/kernel/Makefile
@@ -12,7 +12,7 @@ extra-y	:= head.o vmlinux.lds
 obj-y := entry.o efi.o efi_stub.o gate-data.o fsys.o ia64_ksyms.o irq.o irq_ia64.o	\
 	 irq_lsapic.o ivt.o machvec.o pal.o patch.o process.o perfmon.o ptrace.o sal.o		\
 	 salinfo.o setup.o signal.o sys_ia64.o time.o traps.o unaligned.o \
-	 unwind.o mca.o mca_asm.o topology.o dma-mapping.o
+	 unwind.o mca.o mca_asm.o topology.o dma-mapping.o syscall.o
 
 obj-$(CONFIG_ACPI)		+= acpi.o acpi-ext.o
 obj-$(CONFIG_IA64_BRL_EMU)	+= brl_emu.o
diff --git a/arch/ia64/kernel/entry.S b/arch/ia64/kernel/entry.S
index 68362b3..cbdc2a1 100644
--- a/arch/ia64/kernel/entry.S
+++ b/arch/ia64/kernel/entry.S
@@ -1426,335 +1426,3 @@ END(ftrace_stub)
 
 #endif /* CONFIG_FUNCTION_TRACER */
 
-	.rodata
-	.align 8
-	.globl sys_call_table
-sys_call_table:
-	data8 sys_ni_syscall		//  This must be sys_ni_syscall!  See ivt.S.
-	data8 sys_exit				// 1025
-	data8 sys_read
-	data8 sys_write
-	data8 sys_open
-	data8 sys_close
-	data8 sys_creat				// 1030
-	data8 sys_link
-	data8 sys_unlink
-	data8 ia64_execve
-	data8 sys_chdir
-	data8 sys_fchdir			// 1035
-	data8 sys_utimes
-	data8 sys_mknod
-	data8 sys_chmod
-	data8 sys_chown
-	data8 sys_lseek				// 1040
-	data8 sys_getpid
-	data8 sys_getppid
-	data8 sys_mount
-	data8 sys_umount
-	data8 sys_setuid			// 1045
-	data8 sys_getuid
-	data8 sys_geteuid
-	data8 sys_ptrace
-	data8 sys_access
-	data8 sys_sync				// 1050
-	data8 sys_fsync
-	data8 sys_fdatasync
-	data8 sys_kill
-	data8 sys_rename
-	data8 sys_mkdir				// 1055
-	data8 sys_rmdir
-	data8 sys_dup
-	data8 sys_ia64_pipe
-	data8 sys_times
-	data8 ia64_brk				// 1060
-	data8 sys_setgid
-	data8 sys_getgid
-	data8 sys_getegid
-	data8 sys_acct
-	data8 sys_ioctl				// 1065
-	data8 sys_fcntl
-	data8 sys_umask
-	data8 sys_chroot
-	data8 sys_ustat
-	data8 sys_dup2				// 1070
-	data8 sys_setreuid
-	data8 sys_setregid
-	data8 sys_getresuid
-	data8 sys_setresuid
-	data8 sys_getresgid			// 1075
-	data8 sys_setresgid
-	data8 sys_getgroups
-	data8 sys_setgroups
-	data8 sys_getpgid
-	data8 sys_setpgid			// 1080
-	data8 sys_setsid
-	data8 sys_getsid
-	data8 sys_sethostname
-	data8 sys_setrlimit
-	data8 sys_getrlimit			// 1085
-	data8 sys_getrusage
-	data8 sys_gettimeofday
-	data8 sys_settimeofday
-	data8 sys_select
-	data8 sys_poll				// 1090
-	data8 sys_symlink
-	data8 sys_readlink
-	data8 sys_uselib
-	data8 sys_swapon
-	data8 sys_swapoff			// 1095
-	data8 sys_reboot
-	data8 sys_truncate
-	data8 sys_ftruncate
-	data8 sys_fchmod
-	data8 sys_fchown			// 1100
-	data8 ia64_getpriority
-	data8 sys_setpriority
-	data8 sys_statfs
-	data8 sys_fstatfs
-	data8 sys_gettid			// 1105
-	data8 sys_semget
-	data8 sys_semop
-	data8 sys_semctl
-	data8 sys_msgget
-	data8 sys_msgsnd			// 1110
-	data8 sys_msgrcv
-	data8 sys_msgctl
-	data8 sys_shmget
-	data8 sys_shmat
-	data8 sys_shmdt				// 1115
-	data8 sys_shmctl
-	data8 sys_syslog
-	data8 sys_setitimer
-	data8 sys_getitimer
-	data8 sys_ni_syscall			// 1120		/* was: ia64_oldstat */
-	data8 sys_ni_syscall					/* was: ia64_oldlstat */
-	data8 sys_ni_syscall					/* was: ia64_oldfstat */
-	data8 sys_vhangup
-	data8 sys_lchown
-	data8 sys_remap_file_pages		// 1125
-	data8 sys_wait4
-	data8 sys_sysinfo
-	data8 sys_clone
-	data8 sys_setdomainname
-	data8 sys_newuname			// 1130
-	data8 sys_adjtimex
-	data8 sys_ni_syscall					/* was: ia64_create_module */
-	data8 sys_init_module
-	data8 sys_delete_module
-	data8 sys_ni_syscall			// 1135		/* was: sys_get_kernel_syms */
-	data8 sys_ni_syscall					/* was: sys_query_module */
-	data8 sys_quotactl
-	data8 sys_bdflush
-	data8 sys_sysfs
-	data8 sys_personality			// 1140
-	data8 sys_ni_syscall		// sys_afs_syscall
-	data8 sys_setfsuid
-	data8 sys_setfsgid
-	data8 sys_getdents
-	data8 sys_flock				// 1145
-	data8 sys_readv
-	data8 sys_writev
-	data8 sys_pread64
-	data8 sys_pwrite64
-	data8 sys_sysctl			// 1150
-	data8 sys_mmap
-	data8 sys_munmap
-	data8 sys_mlock
-	data8 sys_mlockall
-	data8 sys_mprotect			// 1155
-	data8 ia64_mremap
-	data8 sys_msync
-	data8 sys_munlock
-	data8 sys_munlockall
-	data8 sys_sched_getparam		// 1160
-	data8 sys_sched_setparam
-	data8 sys_sched_getscheduler
-	data8 sys_sched_setscheduler
-	data8 sys_sched_yield
-	data8 sys_sched_get_priority_max	// 1165
-	data8 sys_sched_get_priority_min
-	data8 sys_sched_rr_get_interval
-	data8 sys_nanosleep
-	data8 sys_ni_syscall			// old nfsservctl
-	data8 sys_prctl				// 1170
-	data8 sys_getpagesize
-	data8 sys_mmap2
-	data8 sys_pciconfig_read
-	data8 sys_pciconfig_write
-	data8 sys_perfmonctl			// 1175
-	data8 sys_sigaltstack
-	data8 sys_rt_sigaction
-	data8 sys_rt_sigpending
-	data8 sys_rt_sigprocmask
-	data8 sys_rt_sigqueueinfo		// 1180
-	data8 sys_rt_sigreturn
-	data8 sys_rt_sigsuspend
-	data8 sys_rt_sigtimedwait
-	data8 sys_getcwd
-	data8 sys_capget			// 1185
-	data8 sys_capset
-	data8 sys_sendfile64
-	data8 sys_ni_syscall		// sys_getpmsg (STREAMS)
-	data8 sys_ni_syscall		// sys_putpmsg (STREAMS)
-	data8 sys_socket			// 1190
-	data8 sys_bind
-	data8 sys_connect
-	data8 sys_listen
-	data8 sys_accept
-	data8 sys_getsockname			// 1195
-	data8 sys_getpeername
-	data8 sys_socketpair
-	data8 sys_send
-	data8 sys_sendto
-	data8 sys_recv				// 1200
-	data8 sys_recvfrom
-	data8 sys_shutdown
-	data8 sys_setsockopt
-	data8 sys_getsockopt
-	data8 sys_sendmsg			// 1205
-	data8 sys_recvmsg
-	data8 sys_pivot_root
-	data8 sys_mincore
-	data8 sys_madvise
-	data8 sys_newstat			// 1210
-	data8 sys_newlstat
-	data8 sys_newfstat
-	data8 sys_clone2
-	data8 sys_getdents64
-	data8 sys_getunwind			// 1215
-	data8 sys_readahead
-	data8 sys_setxattr
-	data8 sys_lsetxattr
-	data8 sys_fsetxattr
-	data8 sys_getxattr			// 1220
-	data8 sys_lgetxattr
-	data8 sys_fgetxattr
-	data8 sys_listxattr
-	data8 sys_llistxattr
-	data8 sys_flistxattr			// 1225
-	data8 sys_removexattr
-	data8 sys_lremovexattr
-	data8 sys_fremovexattr
-	data8 sys_tkill
-	data8 sys_futex				// 1230
-	data8 sys_sched_setaffinity
-	data8 sys_sched_getaffinity
-	data8 sys_set_tid_address
-	data8 sys_fadvise64_64
-	data8 sys_tgkill 			// 1235
-	data8 sys_exit_group
-	data8 sys_lookup_dcookie
-	data8 sys_io_setup
-	data8 sys_io_destroy
-	data8 sys_io_getevents			// 1240
-	data8 sys_io_submit
-	data8 sys_io_cancel
-	data8 sys_epoll_create
-	data8 sys_epoll_ctl
-	data8 sys_epoll_wait			// 1245
-	data8 sys_restart_syscall
-	data8 sys_semtimedop
-	data8 sys_timer_create
-	data8 sys_timer_settime
-	data8 sys_timer_gettime			// 1250
-	data8 sys_timer_getoverrun
-	data8 sys_timer_delete
-	data8 sys_clock_settime
-	data8 sys_clock_gettime
-	data8 sys_clock_getres			// 1255
-	data8 sys_clock_nanosleep
-	data8 sys_fstatfs64
-	data8 sys_statfs64
-	data8 sys_mbind
-	data8 sys_get_mempolicy			// 1260
-	data8 sys_set_mempolicy
-	data8 sys_mq_open
-	data8 sys_mq_unlink
-	data8 sys_mq_timedsend
-	data8 sys_mq_timedreceive		// 1265
-	data8 sys_mq_notify
-	data8 sys_mq_getsetattr
-	data8 sys_kexec_load
-	data8 sys_ni_syscall			// reserved for vserver
-	data8 sys_waitid			// 1270
-	data8 sys_add_key
-	data8 sys_request_key
-	data8 sys_keyctl
-	data8 sys_ioprio_set
-	data8 sys_ioprio_get			// 1275
-	data8 sys_move_pages
-	data8 sys_inotify_init
-	data8 sys_inotify_add_watch
-	data8 sys_inotify_rm_watch
-	data8 sys_migrate_pages			// 1280
-	data8 sys_openat
-	data8 sys_mkdirat
-	data8 sys_mknodat
-	data8 sys_fchownat
-	data8 sys_futimesat			// 1285
-	data8 sys_newfstatat
-	data8 sys_unlinkat
-	data8 sys_renameat
-	data8 sys_linkat
-	data8 sys_symlinkat			// 1290
-	data8 sys_readlinkat
-	data8 sys_fchmodat
-	data8 sys_faccessat
-	data8 sys_pselect6
-	data8 sys_ppoll				// 1295
-	data8 sys_unshare
-	data8 sys_splice
-	data8 sys_set_robust_list
-	data8 sys_get_robust_list
-	data8 sys_sync_file_range		// 1300
-	data8 sys_tee
-	data8 sys_vmsplice
-	data8 sys_fallocate
-	data8 sys_getcpu
-	data8 sys_epoll_pwait			// 1305
-	data8 sys_utimensat
-	data8 sys_signalfd
-	data8 sys_ni_syscall
-	data8 sys_eventfd
-	data8 sys_timerfd_create		// 1310
-	data8 sys_timerfd_settime
-	data8 sys_timerfd_gettime
-	data8 sys_signalfd4
-	data8 sys_eventfd2
-	data8 sys_epoll_create1			// 1315
-	data8 sys_dup3
-	data8 sys_pipe2
-	data8 sys_inotify_init1
-	data8 sys_preadv
-	data8 sys_pwritev			// 1320
-	data8 sys_rt_tgsigqueueinfo
-	data8 sys_recvmmsg
-	data8 sys_fanotify_init
-	data8 sys_fanotify_mark
-	data8 sys_prlimit64			// 1325
-	data8 sys_name_to_handle_at
-	data8 sys_open_by_handle_at
-	data8 sys_clock_adjtime
-	data8 sys_syncfs
-	data8 sys_setns				// 1330
-	data8 sys_sendmmsg
-	data8 sys_process_vm_readv
-	data8 sys_process_vm_writev
-	data8 sys_accept4
-	data8 sys_finit_module			// 1335
-	data8 sys_sched_setattr
-	data8 sys_sched_getattr
-	data8 sys_renameat2
-	data8 sys_getrandom
-	data8 sys_memfd_create			// 1340
-	data8 sys_bpf
-	data8 sys_execveat
-	data8 sys_userfaultfd
-	data8 sys_membarrier
-	data8 sys_kcmp				// 1345
-	data8 sys_mlock2
-	data8 sys_copy_file_range
-	data8 sys_preadv2
-	data8 sys_pwritev2
-
-	.org sys_call_table + 8*NR_syscalls	// guard against failures to increase NR_syscalls
diff --git a/arch/ia64/kernel/syscall.S b/arch/ia64/kernel/syscall.S
new file mode 100644
index 0000000..6d747d1
--- /dev/null
+++ b/arch/ia64/kernel/syscall.S
@@ -0,0 +1,11 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+
+#define __SYSCALL(nr, entry, nargs) data8 entry
+
+.rodata
+.align 8
+.globl sys_call_table
+sys_call_table:
+#include <asm/syscall_64.h>
+	
+.org sys_call_table + 8*NR_syscalls
diff --git a/arch/ia64/kernel/syscalls/Makefile b/arch/ia64/kernel/syscalls/Makefile
new file mode 100644
index 0000000..7916ea8
--- /dev/null
+++ b/arch/ia64/kernel/syscalls/Makefile
@@ -0,0 +1,37 @@
+# SPDX-License-Identifier: GPL-2.0
+out := arch/$(SRCARCH)/include/generated/asm
+uapi := arch/$(SRCARCH)/include/generated/uapi/asm
+
+_dummy := $(shell [ -d '$(uapi)' ] || mkdir -p '$(uapi)') \
+	  $(shell [ -d '$(out)' ] || mkdir -p '$(out)')
+
+syscall := $(srctree)/$(src)/syscall.tbl
+
+syshdr := $(srctree)/$(src)/syscallhdr.sh
+systbl := $(srctree)/$(src)/syscalltbl.sh
+
+quiet_cmd_syshdr = SYSHDR  $@
+      cmd_syshdr = $(CONFIG_SHELL) '$(syshdr)' '$<' '$@'  \
+		   '$(syshdr_abi_$(basetarget))'          \
+		   '$(syshdr_pfx_$(basetarget))'          \
+		   '$(syshdr_offset_$(basetarget))'
+
+quiet_cmd_systbl = SYSTBL  $@
+      cmd_systbl = $(CONFIG_SHELL) '$(systbl)' '$<' '$@'  \
+		   '$(systbl_abi_$(basetarget))'
+
+$(uapi)/unistd_64.h: $(syscall) $(syshdr)
+	$(call if_changed,syshdr)
+
+$(out)/syscall_64.h: $(syscall) $(systbl)
+	$(call if_changed,systbl)
+
+uapisyshdr-y			+= unistd_64.h
+syshdr-y			+= syscall_64.h
+
+targets	+= $(uapisyshdr-y) $(syshdr-y)
+
+PHONY += all
+all: $(addprefix $(uapi)/,$(uapisyshdr-y))
+all: $(addprefix $(out)/,$(syshdr-y))
+	@:
diff --git a/arch/ia64/kernel/syscalls/syscall.tbl b/arch/ia64/kernel/syscalls/syscall.tbl
new file mode 100644
index 0000000..ad410a5
--- /dev/null
+++ b/arch/ia64/kernel/syscalls/syscall.tbl
@@ -0,0 +1,334 @@
+#
+# 64-bit system call numbers and entry vectors
+#
+# The format is:
+# <number> <abi> <name> <entry point>
+#
+# The abi is always "common" for this file.
+#
+1024    common  ni_syscall              sys_ni_syscall
+1025    common  exit                    sys_exit
+1026    common  read                    sys_read
+1027    common  write                   sys_write
+1028    common  open                    sys_open
+1029    common  close                   sys_close
+1030    common  creat                   sys_creat
+1031    common  link                    sys_link
+1032    common  unlink                  sys_unlink
+1033    common  execve                  ia64_execve
+1034    common  chdir                   sys_chdir
+1035    common  fchdir                  sys_fchdir
+1036    common  utimes                  sys_utimes
+1037    common  mknod                   sys_mknod
+1038    common  chmod                   sys_chmod
+1039    common  chown                   sys_chown
+1040    common  lseek                   sys_lseek
+1041    common  getpid                  sys_getpid
+1042    common  getppid                 sys_getppid
+1043    common  mount                   sys_mount
+1044    common  umount                  sys_umount
+1045    common  setuid                  sys_setuid
+1046    common  getuid                  sys_getuid
+1047    common  geteuid                 sys_geteuid
+1048    common  ptrace                  sys_ptrace
+1049    common  access                  sys_access
+1050    common  sync                    sys_sync
+1051    common  fsync                   sys_fsync
+1052    common  fdatasync               sys_fdatasync
+1053    common  kill                    sys_kill
+1054    common  rename                  sys_rename
+1055    common  mkdir                   sys_mkdir
+1056    common  rmdir                   sys_rmdir
+1057    common  dup                     sys_dup
+1058    common  pipe                    sys_ia64_pipe
+1059    common  times                   sys_times
+1060    common  brk                     ia64_brk
+1061    common  setgid                  sys_setgid
+1062    common  getgid                  sys_getgid
+1063    common  getegid                 sys_getegid
+1064    common  acct                    sys_acct
+1065    common  ioctl                   sys_ioctl
+1066    common  fcntl                   sys_fcntl
+1067    common  umask                   sys_umask
+1068    common  chroot                  sys_chroot
+1069    common  ustat                   sys_ustat
+1070    common  dup2                    sys_dup2
+1071    common  setreuid                sys_setreuid
+1072    common  setregid                sys_setregid
+1073    common  getresuid               sys_getresuid
+1074    common  setresuid               sys_setresuid
+1075    common  getresgid               sys_getresgid
+1076    common  setresgid               sys_setresgid
+1077    common  getgroups               sys_getgroups
+1078    common  setgroups               sys_setgroups
+1079    common  getpgid                 sys_getpgid
+1080    common  setpgid                 sys_setpgid
+1081    common  setsid                  sys_setsid
+1082    common  getsid                  sys_getsid
+1083    common  sethostname             sys_sethostname
+1084    common  setrlimit               sys_setrlimit
+1085    common  getrlimit               sys_getrlimit
+1086    common  getrusage               sys_getrusage
+1087    common  gettimeofday            sys_gettimeofday
+1088    common  settimeofday            sys_settimeofday
+1089    common  select                  sys_select
+1090    common  poll                    sys_poll
+1091    common  symlink                 sys_symlink
+1092    common  readlink                sys_readlink
+1093    common  uselib                  sys_uselib
+1094    common  swapon                  sys_swapon
+1095    common  swapoff                 sys_swapoff
+1096    common  reboot                  sys_reboot
+1097    common  truncate                sys_truncate
+1098    common  ftruncate               sys_ftruncate
+1099    common  fchmod                  sys_fchmod
+1100    common  fchown                  sys_fchown
+1101    common  getpriority             ia64_getpriority
+1102    common  setpriority             sys_setpriority
+1103    common  statfs                  sys_statfs
+1104    common  fstatfs                 sys_fstatfs
+1105    common  gettid                  sys_gettid
+1106    common  semget                  sys_semget
+1107    common  semop                   sys_semop
+1108    common  semctl                  sys_semctl
+1109    common  msgget                  sys_msgget
+1110    common  msgsnd                  sys_msgsnd
+1111    common  msgrcv                  sys_msgrcv
+1112    common  msgctl                  sys_msgctl
+1113    common  shmget                  sys_shmget
+1114    common  shmat                   sys_shmat
+1115    common  shmdt                   sys_shmdt
+1116    common  shmctl                  sys_shmctl
+1117    common  syslog                  sys_syslog
+1118    common  setitimer               sys_setitimer
+1119    common  getitimer               sys_getitimer
+1120	common	old_stat		sys_ni_syscall
+1121	common	old_lstat		sys_ni_syscall
+1122	common	old_fstat		sys_ni_syscall
+1123    common  vhangup                 sys_vhangup
+1124    common  lchown                  sys_lchown
+1125    common  remap_file_pages        sys_remap_file_pages
+1126    common  wait4                   sys_wait4
+1127    common  sysinfo                 sys_sysinfo
+1128    common  clone                   sys_clone
+1129    common  setdomainname           sys_setdomainname
+1130    common  uname                   sys_newuname
+1131    common  adjtimex                sys_adjtimex
+1132	common	create_module		sys_ni_syscall
+1133    common  init_module             sys_init_module
+1134    common  delete_module           sys_delete_module
+1135	common	get_kernel_syms		sys_ni_syscall
+1136	common	query_module		sys_ni_syscall
+1137    common  quotactl                sys_quotactl
+1138    common  bdflush                 sys_bdflush
+1139    common  sysfs                   sys_sysfs
+1140    common  personality             sys_personality
+1141    common  afs_syscall             sys_ni_syscall
+1142    common  setfsuid                sys_setfsuid
+1143    common  setfsgid                sys_setfsgid
+1144    common  getdents                sys_getdents
+1145    common  flock                   sys_flock
+1146    common  readv                   sys_readv
+1147    common  writev                  sys_writev
+1148    common  pread64                 sys_pread64
+1149    common  pwrite64                sys_pwrite64
+1150    common  _sysctl                 sys_sysctl
+1151    common  mmap                    sys_mmap
+1152    common  munmap                  sys_munmap
+1153    common  mlock                   sys_mlock
+1154    common  mlockall                sys_mlockall
+1155    common  mprotect                sys_mprotect
+1156    common  mremap                  ia64_mremap
+1157    common  msync                   sys_msync
+1158    common  munlock                 sys_munlock
+1159    common  munlockall              sys_munlockall
+1160    common  sched_getparam          sys_sched_getparam
+1161    common  sched_setparam          sys_sched_setparam
+1162    common  sched_getscheduler      sys_sched_getscheduler
+1163    common  sched_setscheduler      sys_sched_setscheduler
+1164    common  sched_yield             sys_sched_yield
+1165    common  sched_get_priority_max  sys_sched_get_priority_max
+1166    common  sched_get_priority_min  sys_sched_get_priority_min
+1167    common  sched_rr_get_interval   sys_sched_rr_get_interval
+1168    common  nanosleep               sys_nanosleep
+1169    common  nfsservctl              sys_ni_syscall
+1170    common  prctl                   sys_prctl
+1171    common  getpagesize		sys_getpagesize
+1172    common  mmap_pgoff              sys_mmap_pgoff
+1173    common  pciconfig_read          sys_pciconfig_read
+1174    common  pciconfig_write         sys_pciconfig_write
+1175    common  perfmonctl              sys_perfmonctl
+1176    common  sigaltstack             sys_sigaltstack
+1177    common  rt_sigaction            sys_rt_sigaction
+1178    common  rt_sigpending           sys_rt_sigpending
+1179    common  rt_sigprocmask          sys_rt_sigprocmask
+1180    common  rt_sigqueueinfo         sys_rt_sigqueueinfo
+1181    common  rt_sigreturn            sys_rt_sigreturn
+1182    common  rt_sigsuspend           sys_rt_sigsuspend
+1183    common  rt_sigtimedwait         sys_rt_sigtimedwait
+1184    common  getcwd                  sys_getcwd
+1185    common  capget                  sys_capget
+1186    common  capset                  sys_capset
+1187    common  sendfile                sys_sendfile64
+1188    common  getpmsg                 sys_ni_syscall
+1189    common  putpmsg                 sys_ni_syscall
+1190    common  socket                  sys_socket
+1191    common  bind                    sys_bind
+1192    common  connect                 sys_connect
+1193    common  listen                  sys_listen
+1194    common  accept                  sys_accept
+1195    common  getsockname             sys_getsockname
+1196    common  getpeername             sys_getpeername
+1197    common  socketpair              sys_socketpair
+1198    common  send                    sys_send
+1199    common  sendto                  sys_sendto
+1200    common  recv                    sys_recv
+1201    common  recvfrom                sys_recvfrom
+1202    common  shutdown                sys_shutdown
+1203    common  setsockopt              sys_setsockopt
+1204    common  getsockopt              sys_getsockopt
+1205    common  sendmsg                 sys_sendmsg
+1206    common  recvmsg                 sys_recvmsg
+1207    common  pivot_root              sys_pivot_root
+1208    common  mincore                 sys_mincore
+1209    common  madvise                 sys_madvise
+1210    common  stat                    sys_newstat
+1211    common  lstat                   sys_newlstat
+1212    common  fstat                   sys_newfstat
+1213    common  clone2                  sys_clone2
+1214    common  getdents64              sys_getdents64
+1215    common  getunwind               sys_getunwind
+1216    common  readahead               sys_readahead
+1217    common  setxattr                sys_setxattr
+1218    common  lsetxattr               sys_lsetxattr
+1219    common  fsetxattr               sys_fsetxattr
+1220    common  getxattr                sys_getxattr
+1221    common  lgetxattr               sys_lgetxattr
+1222    common  fgetxattr               sys_fgetxattr
+1223    common  listxattr               sys_listxattr
+1224    common  llistxattr              sys_llistxattr
+1225    common  flistxattr              sys_flistxattr
+1226    common  removexattr             sys_removexattr
+1227    common  lremovexattr            sys_lremovexattr
+1228    common  fremovexattr            sys_fremovexattr
+1229    common  tkill                   sys_tkill
+1230    common  futex                   sys_futex
+1231    common  sched_setaffinity       sys_sched_setaffinity
+1232    common  sched_getaffinity       sys_sched_getaffinity
+1233    common  set_tid_address         sys_set_tid_address
+1234    common  fadvise64               sys_fadvise64_64
+1235    common  tgkill                  sys_tgkill
+1236    common  exit_group              sys_exit_group
+1237    common  lookup_dcookie          sys_lookup_dcookie
+1238    common  io_setup                sys_io_setup
+1239    common  io_destroy              sys_io_destroy
+1240    common  io_getevents            sys_io_getevents
+1241    common  io_submit               sys_io_submit
+1242    common  io_cancel               sys_io_cancel
+1243    common  epoll_create            sys_epoll_create
+1244    common  epoll_ctl               sys_epoll_ctl
+1245    common  epoll_wait              sys_epoll_wait
+1246    common  restart_syscall         sys_restart_syscall
+1247    common  semtimedop              sys_semtimedop
+1248    common  timer_create            sys_timer_create
+1249    common  timer_settime           sys_timer_settime
+1250    common  timer_gettime           sys_timer_gettime
+1251    common  timer_getoverrun        sys_timer_getoverrun
+1252    common  timer_delete            sys_timer_delete
+1253    common  clock_settime           sys_clock_settime
+1254    common  clock_gettime           sys_clock_gettime
+1255    common  clock_getres            sys_clock_getres
+1256    common  clock_nanosleep         sys_clock_nanosleep
+1257    common  fstatfs64               sys_fstatfs64
+1258    common  statfs64                sys_statfs64
+1259    common  mbind                   sys_mbind
+1260    common  get_mempolicy           sys_get_mempolicy
+1261    common  set_mempolicy           sys_set_mempolicy
+1262    common  mq_open                 sys_mq_open
+1263    common  mq_unlink               sys_mq_unlink
+1264    common  mq_timedsend            sys_mq_timedsend
+1265    common  mq_timedreceive         sys_mq_timedreceive
+1266    common  mq_notify               sys_mq_notify
+1267    common  mq_getsetattr           sys_mq_getsetattr
+1268    common  kexec_load              sys_kexec_load
+1269    common  vserver                 sys_ni_syscall
+1270    common  waitid                  sys_waitid
+1271    common  add_key                 sys_add_key
+1272    common  request_key             sys_request_key
+1273    common  keyctl                  sys_keyctl
+1274    common  ioprio_set              sys_ioprio_set
+1275    common  ioprio_get              sys_ioprio_get
+1276    common  move_pages              sys_move_pages
+1277    common  inotify_init            sys_inotify_init
+1278    common  inotify_add_watch       sys_inotify_add_watch
+1279    common  inotify_rm_watch        sys_inotify_rm_watch
+1280    common  migrate_pages           sys_migrate_pages
+1281    common  openat                  sys_openat
+1282    common  mkdirat                 sys_mkdirat
+1283    common  mknodat                 sys_mknodat
+1284    common  fchownat                sys_fchownat
+1285    common  futimesat               sys_futimesat
+1286    common  newfstatat              sys_newfstatat
+1287    common  unlinkat                sys_unlinkat
+1288    common  renameat                sys_renameat
+1289    common  linkat                  sys_linkat
+1290    common  symlinkat               sys_symlinkat
+1291    common  readlinkat              sys_readlinkat
+1292    common  fchmodat                sys_fchmodat
+1293    common  faccessat               sys_faccessat
+1294    common  pselect6                sys_pselect6
+1295    common  ppoll                   sys_ppoll
+1296    common  unshare                 sys_unshare
+1297    common  splice                  sys_splice
+1298    common  set_robust_list         sys_set_robust_list
+1299    common  get_robust_list         sys_get_robust_list
+1300    common  sync_file_range         sys_sync_file_range
+1301    common  tee                     sys_tee
+1302    common  vmsplice                sys_vmsplice
+1303    common  fallocate               sys_fallocate
+1304    common  getcpu                  sys_getcpu
+1305    common  epoll_pwait             sys_epoll_pwait
+1306    common  utimensat               sys_utimensat
+1307    common  signalfd                sys_signalfd
+1308    common  timerfd                 sys_ni_syscall
+1309    common  eventfd                 sys_eventfd
+1310    common  timerfd_create          sys_timerfd_create
+1311    common  timerfd_settime         sys_timerfd_settime
+1312    common  timerfd_gettime         sys_timerfd_gettime
+1313    common  signalfd4               sys_signalfd4
+1314    common  eventfd2                sys_eventfd2
+1315    common  epoll_create1           sys_epoll_create1
+1316    common  dup3                    sys_dup3
+1317    common  pipe2                   sys_pipe2
+1318    common  inotify_init1           sys_inotify_init1
+1319    common  preadv                  sys_preadv
+1320    common  pwritev                 sys_pwritev
+1321    common  rt_tgsigqueueinfo       sys_rt_tgsigqueueinfo
+1322    common  recvmmsg                sys_recvmmsg
+1323    common  fanotify_init           sys_fanotify_init
+1324    common  fanotify_mark           sys_fanotify_mark
+1325    common  prlimit64               sys_prlimit64
+1326    common  name_to_handle_at       sys_name_to_handle_at
+1327    common  open_by_handle_at       sys_open_by_handle_at
+1328    common  clock_adjtime           sys_clock_adjtime
+1329    common  syncfs                  sys_syncfs
+1330    common  setns                   sys_setns
+1331    common  sendmmsg                sys_sendmmsg
+1332    common  process_vm_readv        sys_process_vm_readv
+1333    common  process_vm_writev       sys_process_vm_writev
+1334    common  accept4                 sys_accept4
+1335    common  finit_module            sys_finit_module
+1336    common  sched_setattr           sys_sched_setattr
+1337    common  sched_getattr           sys_sched_getattr
+1338    common  renameat2               sys_renameat2
+1339    common  getrandom               sys_getrandom
+1340    common  memfd_create            sys_memfd_create
+1341    common  bpf                     sys_bpf
+1342    common  execveat                sys_execveat
+1343    common  userfaultfd             sys_userfaultfd
+1344    common  membarrier              sys_membarrier
+1345    common  kcmp                    sys_kcmp
+1346    common  mlock2                  sys_mlock2
+1347    common  copy_file_range         sys_copy_file_range
+1348    common  preadv2                 sys_preadv2
+1349    common  pwritev2                sys_pwritev2
diff --git a/arch/ia64/kernel/syscalls/syscallhdr.sh b/arch/ia64/kernel/syscalls/syscallhdr.sh
new file mode 100644
index 0000000..5f8d7c0
--- /dev/null
+++ b/arch/ia64/kernel/syscalls/syscallhdr.sh
@@ -0,0 +1,32 @@
+#!/bin/sh
+# SPDX-License-Identifier: GPL-2.0
+
+in="$1"
+out="$2"
+my_abis=`echo "($3)" | tr ',' '|'`
+prefix="$4"
+offset="$5"
+
+fileguard=_UAPI_ASM_IA64_`basename "$out" | sed \
+    -e 'y/abcdefghijklmnopqrstuvwxyz/ABCDEFGHIJKLMNOPQRSTUVWXYZ/' \
+    -e 's/[^A-Z0-9_]/_/g' -e 's/__/_/g'`
+grep -E "^[0-9A-Fa-fXx]+[[:space:]]+${my_abis}" "$in" | sort -n | (
+    echo "#ifndef ${fileguard}"
+    echo "#define ${fileguard}"
+    echo ""
+    echo "#include <asm/break.h>"
+    echo ""
+    echo -e "#define __BREAK_SYSCALL\t__IA64_BREAK_SYSCALL"
+    echo ""
+
+    while read nr abi name entry ; do
+	if [ -z "$offset" ]; then
+	    echo -e "#define __NR_${prefix}${name}\t$nr"
+	else
+	    echo -e "#define __NR_${prefix}${name}\t($offset + $nr)"
+	fi
+    done
+
+    echo ""
+    echo "#endif /* ${fileguard} */"
+) > "$out"
diff --git a/arch/ia64/kernel/syscalls/syscalltbl.sh b/arch/ia64/kernel/syscalls/syscalltbl.sh
new file mode 100644
index 0000000..a7c1c17
--- /dev/null
+++ b/arch/ia64/kernel/syscalls/syscalltbl.sh
@@ -0,0 +1,29 @@
+#!/bin/sh
+# SPDX-License-Identifier: GPL-2.0
+
+in="$1"
+out="$2"
+my_abi="$3"
+
+emit() {
+    nxt="$1"
+    nr="$2"
+    entry="$3"
+
+    while [ $nxt -lt $nr ]; do
+	echo "__SYSCALL($nxt, sys_ni_syscall, )"
+	let nxt=nxt+1
+    done
+
+    echo "__SYSCALL($nr, $entry, )"
+}
+
+
+grep '^[0-9]' "$in" | sort -n | (
+    nxt=1024
+    while read nr abi name entry ; do
+	emit $nxt $nr $entry
+	nxt=$nr
+        let nxt=nxt+1
+    done
+) > "$out"
-- 
1.9.1

